modules/core/src/main/scala/org/tessellation/domain/snapshot/programs/Download.scala

```diff
@@ -0,0 +1,317 @@
+            }
+        }
+
+    def fetchSnapshot(hash: Option[Hash], ordinal: SnapshotOrdinal): F[Signed[GlobalIncrementalSnapshot]] =
+      clusterStorage.getPeers
+        .map(_.toList)
+        .flatMap(Random[F].shuffleList)
+        .flatTap { _ =>
+          logger.info(s"Downloading snapshot hash=${hash.show}, ordinal=${ordinal.show}")
+        }
+        .flatMap { peers =>
+          type Success = Signed[GlobalIncrementalSnapshot]
+          type Result = Option[Success]
+          type Agg = (List[Peer], Result)
+
+          (peers, none[Success]).tailRecM[F, Result] {
+            case (Nil, snapshot) => snapshot.asRight[Agg].pure[F]
+            case (peer :: tail, _) =>
+              p2pClient.globalSnapshot
+                .get(ordinal)
+                .run(peer)
+                .flatMap(_.toHashed[F])
+                .map(_.some)
+                .handleError(_ => none[Hashed[GlobalIncrementalSnapshot]])
+                .map {
+                  case Some(snapshot) if hash.forall(_ === snapshot.hash) => snapshot.signed.some.asRight[Agg]
+                  case _                                                  => (tail, none[Success]).asLeft[Result]
+                }
+          }
+        }
+        .flatMap {
+          case Some(snapshot) => snapshot.pure[F]
+          case _              => CannotFetchSnapshot.raiseError[F, Signed[GlobalIncrementalSnapshot]]
+        }
+
+    def fetchGenesis(ordinal: SnapshotOrdinal): F[Signed[GlobalSnapshot]] =
+      clusterStorage.getPeers
+        .map(_.toList)
+        .flatMap(Random[F].shuffleList)
+        .flatTap { _ =>
+          logger.info(s"Downloading genesis snapshot ordinal=${ordinal}")
+        }
+        .flatMap { peers =>
+          type Success = Signed[GlobalSnapshot]
+          type Agg = (List[Peer], Option[Signed[GlobalSnapshot]])
+          type Result = Option[Success]
+
+          (peers, none[Success]).tailRecM[F, Result] {
+            case (Nil, snapshot) => snapshot.asRight[Agg].pure[F]
+            case (peer :: tail, _) =>
+              p2pClient.globalSnapshot
+                .getFull(ordinal)
+                .run(peer)
+                .flatMap(_.toHashed[F])
+                .map(_.some)
+                .handleError(_ => none[Hashed[GlobalSnapshot]])
+                .map {
+                  case Some(snapshot) => snapshot.signed.some.asRight[Agg]
+                  case _              => (tail, none[Success]).asLeft[Result]
+                }
+          }
+        }
+        .flatMap {
+          case Some(snapshot) => snapshot.pure[F]
+          case _              => CannotFetchGenesisSnapshot.raiseError[F, Signed[GlobalSnapshot]]
+        }
+  }
+
+  case object HashAndOrdinalMismatch extends NoStackTrace
+  case object CannotFetchSnapshot extends NoStackTrace
+  case object CannotFetchGenesisSnapshot extends NoStackTrace
+  case object InvalidChain extends NoStackTrace
+  case object UnexpectedState extends NoStackTrace
```

Comment:
For that particular case it is probably not needed but we often introduce a sealed trait like

```scala
sealed trait DownloadError extends NoStackTrace
```